core/base-service/coalesce-badge.js

```diff
@@ -149,9 +150,16 @@ export default function coalesceBadge(
     logoPosition = coalesce(overrideLogoPosition, serviceLogoPosition)
   }
   if (namedLogo) {
```

Comment:
This means this works for simple-icons, the small handful of shields named logos (which are all pretty close to 'square'), but not custom user-supplied base64 logos.
Is there any particular reason why it would be harder or not make sense to make this work for all logos?

Reply:
User-supplied base64 logos might be non-SVG images. We can figure out its aspect ratio with SVG libraries.

This might need another image process library like https://github.com/FFmpeg/FFmpeg. But this will consume more server resources.

Reply:
OK. I think that's a valid reason to not do it. I'd forgotten base64 logo is not even necessarily a SVG.

`logoWidth` is.. awkward. It already exists as an undocumented param. Also it doesn't do what you would expect: https://img.shields.io/badge/build-passing-brightgreen?logo=javascript&logoWidth=100 it is kind of more like `logoHorizontalPadding` than `logoWidth`. Mostly unrelated to this PR - I think we should probably deprecate/remove this at some point, but of course that will break something for someone. This is its own can of worms, but I think I probably don't want to overload `logoWidth`.

I think my preference would be that we stick with the implementation that only applies to named logos for now. Lets deal with custom logos as another issue (if at all). Sorry we are kinda going back and forth on this one, but I don't want to hold up the part of this that we are fairly settled on with this bit of messy scope creep.

Reply:
Right, yes. Interesting, so this would allow you to take

![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=coinbase&logoSize=auto) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=coinbase&logoSize=auto

and make it

![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoWidth=50) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoWidth=50
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto&logoWidth=50) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto&logoWidth=50
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=coinbase&logoSize=auto&logoWidth=50) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=coinbase&logoSize=auto&logoWidth=50

which is potentially useful :thinking:

Reply:
Do you need to update these two options?

- `logoWidth` -> `logoBoundingWidth`
- `logoHeight` -> `logoBoundingHeight`

This will be a breaking change. My suggestion is to write more detailed description for `logoWidth` and `logoHeight`.

```diff
logoWidth
-Default: 14. Same meaning as the query string. Can be overridden by the query string.
+Default: 14. This parameter is used to change the width of the bounding box for the logo.

logoHeight
-Default: 14. Same meaning as the query string. Can be overridden by the query string.
+Default: 14. This parameter is used to change the height of the bounding box for the logo.
```

Reply:
OK. So I've had a bit of play with this, and a bit of a think. I know we're moving around a bit on the implementation, but being able to play with this on a review app is helping and I appreciate the work you're doing on it. I just want to zoom out on this slightly and think about some of the things we're trying to optimise for as a project. It is important to get this right as once we ship some new query params we're basically stuck with them forever.

- **Consistency**: Thinking about why shields was started in the first place, one of the goals was consistency. https://olivierlacan.com/posts/an-open-source-rage-diamond/ In general, as a project, we value providing a small number of good options over loads of configuration options.
- **Good defaults**: We try to give the user good defaults, rather than put the onus on users to configure. A good example of this is the work we did to try and automatically pick a sensible-ish background colour for logos: https://github.com/badges/shields/pull/2833 Make it easy for the user to "fall into the pit of success".
- **Minimal support burden**: We are a small team of volunteers supporting a large userbase. Adding configuration which is complicated or requires a lot of explanation generates support requests and we have limited time to answer them. Even if we provide good docs, lots of people don't read it.

Thinking through what we've got at the moment, I think the implementation is not that intuitive, even if we rename the params. My instinct is that being able to use a combination of `logoSize=auto` and some fixed witdth is useful but I'm not convinced that allowing one user to have `logoSize=auto&logoWidth=50`, another to have `logoSize=auto&logoWidth=60` and another to have `logoSize=auto&logoWidth=70` is that valuable, and it goes against the goal of consistency. This also allows the user to do wierd or undesirable things like https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoHeight=400

So just zooming back out to the original user need..
The problem we are trying to solve here is really: Some logos are wide and thin, and they look bad in the default viewbox. The original problem we were trying to solve was not so much an appetite for limitless customisation.

Given that, I think my suggestion is that we try and limit this to some "presets". Maybe

- nothing/default
- `logoSize=auto`
- `logoSize=wide` (a fixed-width that looks good with wide logos, probably something like `logoSize=auto&logoWidth=40` in the current implementation)

but remove the ability for the user to explicitly set the viewbox numerically, and don't worry too much about user-supplied logos for now.

Thoughts?

Reply:
I think we can limit it's max-width to the full-filled-width.

So the https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoHeight=400 will still be: https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto

So the code will be:

```diff
-logoWidth = (iconSize.width / iconSize.height) * (logoHeight || DEFAULT_LOGO_HEIGHT)
+const autoLogoWidth = (iconSize.width / iconSize.height) * (logoHeight || DEFAULT_LOGO_HEIGHT)
+logoWidth = logoWidth > autoLogoWidth ? autoLogoWidth : logoWidth
```

What do you think?

Reply:
That constraint would help with the 'wierd or undesirable' outputs, but I think the other points still stand.

Reply:
The `logoWidth: 'auto' | 'wide'` doesn't display icons well in all scenarios.

The most wide icon is `comsol`. The logo still looks so small even we set width to 40. So it's hard to find out some preset values for icons.

I think it is still necessary to choose the appropriate width according to their icons and the context of the icon and text content.

![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=comsol&logoSize=auto&logoWidth=40) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=cosmol&logoSize=auto&logoWidth=40
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoWidth=40) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoWidth=40
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto&logoWidth=40) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoWidth=40

The `logoHeight` is also useful, you can understand the icon with 14 zoom levels. The `1` for the minimum size, `7` for the middle size, and `14` for the max size.

![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=comsol&logoSize=auto&logoHeight=7) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=cosmol&logoSize=auto&logoHeight=7
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoHeight=7) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amg&logoSize=auto&logoHeihgt=7
![](https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto&logoHeight=7) https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=amd&logoSize=auto&logoHeihgt=7

This function is convenient for many people to personalize his badge wall, and it is not a complex function for our maintainers to maintain.

Reply:
For anyone confused by the above links https://pr-9191-badges-shields.fly.dev/badge/foo-bar-blue?logo=comsol&logoSize=auto&logoWidth=40 is the right link :smile: 

> Personally I think it is not a complex function to maintain.

I agree it isn't a difficult feature to maintain in the sense that I don't think we would be constantly updating this code over time. I'm sure you'll agree replying to support requests is a form of maintenance though so I'm keen that we find a solution that is intuitive.

---

@calebcartwright  @PyvesB  Either of you got any thoughts on this one?

If the review app has been cleaned up, you can re-deploy it:

![Screenshot at 2023-06-23 19-30-19](https://github.com/badges/shields/assets/6025893/1f853fb2-3dd6-4ae1-a828-cd827c344a00)

Reply:
I've not read through all the details in this PR, and the review app is failing to deploy again. I'll try to circle back to this tomorrow.

Once thing I want to bring back to our attention is https://github.com/badges/shields/issues/3635. There was quite a lot of heated debate at the time about how we had allowed for inconsistent logo positioning and sizing with the popout style. The resolution plan also included removal of the `logoPosition` code at the time, but we apparently didn't do it. We need to be careful if we want to reintroduce new options related to logo appearance.

Reply:
Yes I am feeling echoes of the popout style here and keen find a way to solve this without going down the same route.
I think the fact that `logoWidth` still exists is a remnant of that.

The reason why the review app is not deploying is because of the merge conflicts (side note: I'm going to raise an issue about that). In order to get this deploying again, we'll need to resolve the conflicts. Are you able to have a look at that @LitoMore ? :pray:

Reply:
I have rebased with the latest master. It should be working now. I have no permission to trigger workflows.

Reply:
I also created a PR to adjust the `deploy-review-app` script.

- https://github.com/badges/shields/pull/9317

Reply:
I can revert the exposed `logoWidth` `logoHeight` codes and wait for other users' feedback if we still have reservations about this.

Reply:
> @calebcartwright @PyvesB Either of you got any thoughts on this one?

I confess I probably need to go back and do a thorough read through the motivating request/issue in #8372 and the full commentary. However, my current impression is that we've started drifting from the "how" (the implementation) and are circling back to the "what" and "why"

The way you described this @chris48s above:

> So just zooming back out to the original user need..
The problem we are trying to solve here is really: Some logos are wide and thin, and they look bad in the default viewbox. The original problem we were trying to solve was not so much an appetite for limitless customisation.

Still resonates with me, and that seems like a valid thing we should consider whether we want to address (and if so, be able to articulate why).

As many will likely remember, I'm quite adamantly opposed to infinite customization. However, I don't inherently currently view attempts to address that user need as being contradictory with the goals of the project, nor as falling into that "too much customization" bucket (although I recognize certain implementations/solutions _could_).

I.e. I suspect there's a universe in which we could potentially address the user need that would still align with our goals and self-imposed constraints around where we draw the customization line

Reply:
> I can revert the exposed logoWidth logoHeight codes and wait for other users' feedback if we still have reservations about this.

Yes. An incremental approach is one way forward. We could add the `logoSize=auto` param as a first step and then see where we think we should go next based on user feedback. Adding that doesn't prevent us adding more params later. Once we have added params or settings, removing or changing them is harder: We can't really version a service like we can with a library.